ramips: add support for LW-IP301

koreapyj opened the pull request:
Add support for LW-IP301.

Hardware:
SoC: Mediatek MT7620N (MIPS 580MHz)
RAM: 64MB
Flash: Winbond W25Q64BV (8MB)
WiFi: Mediatek RT5390 Integrated Wireless (2.4GHz)
Ethernet: Mediatek MT7620 Integrated Switch (WAN x1, LAN x2)
UART: [GND, RX, TX] (57600)

Installation:
Connect UART. Square pin is GND.
Select TFTP flashing mode from U-Boot and flash sysupgrade image.

mrkiko commented:
Hi!
thanks for your work!

May you elaborate a little bit on flashing instructions?
Is UART necessary to flash? If so, may you detail the required procedure?

Thanks!

koreapyj commented:
> May you elaborate a little bit on flashing instructions?
> Is UART necessary to flash? If so, may you detail the required procedure?

Updated. May be stock firmware has better way to flash but I lost stock firmware and there is no vendor website or firmware image so currently TFTP with UART is only way to flash. :(

mans0n commented on target/linux/ramips/dts/mt7620n_generic_lw-ip301.dts:
```diff
@@ -0,0 +1,130 @@
+
+&state_default {
+	default {
+		groups = "ephy", "wled", "i2c", "wdt", "pa", "spi refclk";
```
Do you need all of this?

koreapyj replied:
It has some issues about ethernet on my 21.02 port but now it looks like useless. removed.

mans0n replied:
But you'll definitely need some pinctrls for keys and leds...

mans0n replied:
I think you are using only `"i2c", "spi refclk", "ephy"` pins, and that should be added to pinctrl. Please let me know if the board has some other exported pins.
BTW it seems the ethernet leds can be used with hardware trigger... But I won't object if you choose to use software trigger.

mans0n commented:
> Select TFTP flashing mode from U-Boot and flash initramfs image.

I guess we can flash sysupgrade image directly, no?

koreapyj commented:
> > Select TFTP flashing mode from U-Boot and flash initramfs image.
> 
> I guess we can flash sysupgrade image directly, no?

Could be done with sysupgrade. Updated.

mans0n commented:
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#answer_questions

Please answer my questions before marking them as resolved.

koreapyj commented on target/linux/ramips/dts/mt7620n_generic_lw-ip301.dts:
```diff
@@ -0,0 +1,130 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
It works correctly on 50000000. Applied.

koreapyj commented on target/linux/ramips/dts/mt7620n_generic_lw-ip301.dts:
```diff
@@ -0,0 +1,130 @@
+			partition@50000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0x780000>;
```
It just copy & paste stock partition layout. Nothing is there. Extended to full flash.

koreapyj commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -444,6 +444,15 @@ define Device/fon_fon2601
 TARGET_DEVICES += fon_fon2601
 
+define Device/generic_lw-ip301
+  $(Device/uimage-lzma-loader)
```
Same as "state_default" things. removed.

mans0n commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -444,6 +444,15 @@ define Device/fon_fon2601
+  $(Device/uimage-lzma-loader)
+  SOC := mt7620n
+  IMAGE_SIZE := 7872k
+  DEVICE_VENDOR := Generic
```
Umm, actually my first thought was adding "unbranded" only to the boardname strings, not to model/DEVICE_VENDOR.
I've opened a PR #9327 to handle the other unbranded devices, let's decide which way to take after we receive some comments there.

mans0n replied:
It's good to go in this state. 94d4269ec7

fransix commented:
> 
Hello koreapyj, I have tried to build an image for my lw-ip301 board and the wifi does not work, wifi config is empty and at boot I get:
[   10.918843] ieee80211 phy0: rt2x00_set_rt: Info - RT chipset 6352, rev 0500 detected
[   10.934334] ieee80211 phy0: rt2800_init_eeprom: Error - Invalid RF chipset 0x0527 detected
[   10.950827] ieee80211 phy0: rt2x00lib_probe_dev: Error - Failed to allocate device

I tried to add the firmware to the image but I got the same result:
rt2800-pci-firmware.................. Ralink RT28xx/3xxx PCI/SoC firmware

Did the wifi work for you?

mans0n commented:
@fransix I think the problem is in the EEPROM. Could you dump the factory partition? (`hexdump -C /dev/mtd2` | head)

mans0n commented:
Strange, that's not the factory partition. Could you show me the result of `cat /proc/mtd`?

mans0n commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -444,6 +444,14 @@ define Device/fon_fon2601
 endef
 TARGET_DEVICES += fon_fon2601
 
+define Device/unbranded_lw-ip301
```
Sorting.

mans0n commented on target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:
```diff
@@ -109,6 +109,11 @@ engenius,esr600)
 	ucidef_set_led_netdev "wlan5g" "5.0GHz" "blue:wlan5g" "wlan0"
 	ucidef_set_led_netdev "wlan2g" "2.4GHz" "blue:wlan2g" "wlan1"
 	;;
+unbranded,lw-ip301)
```
Sorting.

mans0n commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -176,6 +176,7 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1" "0:wan" "6@eth0"
 		;;
+	unbranded,lw-ip301|\
```
Sorting.

mans0n commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -257,6 +258,7 @@ ramips_setup_macs()
 	bdcom,wap2100-sk|\
 	edimax,ew-7478apc|\
 	fon,fon2601|\
+	generic,lw-ip301|\
```
Duplicate.

mans0n commented:
I mean, does the stock FW have working wifi *now*? I'm asking this because you might not have working wifi anymore if you lost your factory partition.

fransix commented:
It does not work since I flashed an OpenWrt image.  Wifi status is empty

mans0n commented:
Too bad. :( You'll need to get the factory partition data from somewhere for recovery...

I'd like to know if the installation procedure is faulty or not...

fransix commented:
These boards are cheap, I'll order another one and keep the original FW in it (and always load Openwrt into RAM) to make some eeprom comparisons.  All I did on this one is connect to the serial console to access U-Boot options to load a sysupgrade into flash.  It was not possible to transfer an image through a telnet session (and no ssh).  I would supose that the procedure is faulty. My options were extremely limited and I didn't do anything special...

fransix commented:
I received another unit.  I now have a unit with the OEM FW and my initial unit for which I built a sysupgrade and flashed using the bootloader TFTP tool but had wifi problems afterwards (possibly because of wifi calib data erased?)

here is the OEM FW mapping:

W25Q64BV(ef 40170000) (8192 Kbytes)
mtd .name = raspi, .size = 0x00800000 (0M) .erasesize = 0x00000008 (0K) .numeraseregions = 65536
Creating 6 MTD partitions on "raspi":
0x000000000000-0x000000800000 : "ALL"
0x000000000000-0x000000030000 : "Bootloader"
0x000000030000-0x000000040000 : "Config"
0x000000040000-0x000000050000 : "Factory"
0x000000050000-0x000000380000 : "Kernel"
0x000000380000-0x000000400000 : "Fconfig"

and here is my sysupgrade mapping:

[    0.688875] spi-nor spi0.0: s25fl064k (8192 Kbytes)
[    0.698754] 4 fixed-partitions partitions found on MTD device spi0.0
[    0.711432] Creating 4 MTD partitions on "spi0.0":
[    0.720993] 0x000000000000-0x000000030000 : "u-boot"
[    0.733087] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.744905] 0x000000040000-0x000000050000 : "factory"
[    0.758061] 0x000000050000-0x000000800000 : "firmware"
[    0.769646] 2 uimage-fw partitions found on MTD device firmware
[    0.781542] Creating 2 MTD partitions on "firmware":
[    0.791457] 0x000000000000-0x0000001eb06d : "kernel"
[    0.801364] mtd: partition "kernel" doesn't end on an erase/write block -- force read-only
[    0.820715] 0x0000001eb06d-0x0000007b0000 : "rootfs"
[    0.830691] mtd: partition "rootfs" doesn't start on an erase/write block boundary -- force read-only
[    0.850163] mtd: device 5 (rootfs) set to be root filesystem
[    0.862597] 1 squashfs-split partitions found on MTD device rootfs
[    0.875004] 0x00000054e000-0x0000007b0000 : "rootfs_data"

I noticed the OEM bootloader factory TFTP option writes starting at 0x50000.  So I extracted the OEM FW by "cat"ing mtd3 through mtd5 and saved them in binary format via the serial port console.  Then I flashed this file to my openwrt unit still using the OEM bootloader factory TFTP but it didn't boot saying "Bad Magic Number".  

I think the problem is that the factory partitions do not start at the same addresses (OEM: 0x40000 vs sysupgrade:0x50000).  What should I do to build a compatible sysupgrade?

Thanks

fransix commented:
I have received another unit.  I first tried to extract the stock FW by directly cat'ing it to the serial port via stdout but I found out there are always some bytes that are lost using this process (and I'm using an FTDI usb to serial converter, which usually performs well).  Sadly there are no other means to transfer a file going out using the stock FW or the stock bootloader.

So I had no choice to give another try to the installation procedure with this new unit.  I first found out that loading an initramfs build into RAM works but still breaks something which prevents the stock FW to boot properly afterwards.

I then flashed a sysupgrade build onto the new unit and, suprisingly, everything seems to work including wifi.  Here what the MTD looks like:
[    0.676648] spi spi0.0: force spi mode3
[    0.685756] spi-nor spi0.0: s25fl064k (8192 Kbytes)
[    0.695631] 4 fixed-partitions partitions found on MTD device spi0.0
[    0.708300] Creating 4 MTD partitions on "spi0.0":
[    0.717873] 0x000000000000-0x000000030000 : "u-boot"
[    0.729964] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.741770] 0x000000040000-0x000000050000 : "factory"
[    0.754917] 0x000000050000-0x000000800000 : "firmware"
[    0.766505] 2 uimage-fw partitions found on MTD device firmware
[    0.778396] Creating 2 MTD partitions on "firmware":
[    0.788301] 0x000000000000-0x0000001eee05 : "kernel"
[    0.798197] mtd: partition "kernel" doesn't end on an erase/write block -- force read-only
[    0.817510] 0x0000001eee05-0x0000007b0000 : "rootfs"
[    0.827484] mtd: partition "rootfs" doesn't start on an erase/write block boundary -- force read-only
[    0.846944] mtd: device 5 (rootfs) set to be root filesystem
[    0.859361] 1 squashfs-split partitions found on MTD device rootfs
[    0.871759] 0x00000045a000-0x0000007b0000 : "rootfs_data"

The /dev/mtd2 (factory) partition seems okay this time.  I now have been able to dd it to a file and scp it out to my old unit.  I am now trying to make my old unit work properly by rewriting mtd2 but neither
mtd -f write /tmp/mtd2.bin factory
nor
mtd -f write /tmp/mtd2.bin mtd2
work, both reply "Can't open device for writing!"

So in the end the procedure works but I really have no idea as to what happened at my first attempt for the mtd2 (factory) partition to be written to with a kernel image because neither the bootloader tools nor, I think, any sysupgrade should write at that location...

mans0n commented:
> Sadly there are no other means to transfer a file going out using the stock FW or the stock bootloader.

My favorite approach for dumping stock firmware is to boot initramfs into RAM and dd & scp there. Normally just booting initramfs should not affect the flash in any way.

> both reply "Can't open device for writing!" and same for "mtd unlock mtd2".

If you want to write read-only partition then you need to drop the read-only property in the DTS or install `kmod-mtd-rw`.

I also don't understand what happened to your first device.

The `wifi up` log shows some warnings (on which I have no idea) but the wifi seems to be enabled after all, isn't it working properly?

fransix commented:
> The `wifi up` log shows some warnings (on which I have no idea) but the wifi seems to be enabled after all, isn't it working properly?

No, the wifi is not broadcasting any signal...

fransix commented:
GOOD NEWS!
it doesn't seem to work on my new unit BUT I copied the factory partition to my old unit (insmod'ing mtd-rw, thanks) and I DO have a working wifi!    :)    I will now try to find out what is wrong with my new unit.  It looks like I often do something wrong before getting it right

fransix commented:
I confirm again that the procedure works.  I have ordered and received yet another unit.  Before I flashed it, I booted with an initramfs image and extracted the factory stock fw and all other partitions, for whom it would be useful.
[lw-ip301_mtd.zip](https://github.com/openwrt/openwrt/files/8741942/lw-ip301_mtd.zip)

koreapyj committed:
ramips: add support for LW-IP301
Add support for LW-IP301.

Hardware:
SoC: Mediatek MT7620N (MIPS 580MHz)
RAM: 64MB
Flash: Winbond W25Q64BV (8MB)
WiFi: Mediatek RT5390 Integrated Wireless (2.4GHz)
Ethernet: Mediatek MT7620 Integrated Switch (WAN x1, LAN x2)
UART: [GND, RX, TX] (57600)
- At near the reset button.
- Square pin is TX.

Installation:
Connect UART.
Select TFTP flashing mode from U-Boot and flash sysupgrade image.

Signed-off-by: Yoonji Park <koreapyj@dcmys.kr>

koreapyj committed:
uboot
